// @tpluscode/rdf-ns-builders/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.qkdv = exports.qb = exports.prov = exports.prefix = exports.owl = exports.org = exports.og = exports.oa = exports.ma = exports.lvont = exports.locn = exports.ldp = exports.ical = exports.hydra = exports.http = exports.gtfs = exports.gs1 = exports.grddl = exports.gr = exports.gn = exports.gml = exports.geor = exports.geof = exports.geo = exports.frbr = exports.foaf = exports.exif = exports.ebucore = exports.earl = exports.duv = exports.dtype = exports.dqv = exports.dpv = exports.doap = exports.discipline = exports.dcterms = exports.dcmitype = exports.dcat = exports.dcam = exports.dc11 = exports.dbo = exports.dash = exports.ctag = exports.csvw = exports.constant = exports.cnt = exports.cc = exports.bibo = exports.as = exports.acl = void 0;
exports.xsd = exports.xml = exports.xkos = exports.xhv = exports.wgs = exports.wdrs = exports.wdr = exports.vs = exports._void = exports.vcard = exports.vann = exports.vaem = exports.v = exports.unit = exports.time = exports.test = exports.ssn = exports.sou = exports.sosa = exports.skosxl = exports.skos = exports.sioc = exports.shex = exports.sh = exports.sf = exports.sem = exports.sdmx = exports.sd = exports.schema = exports.rss = exports.rr = exports.rif = exports.rico = exports.rev = exports.rdfs = exports.rdfa = exports.rdf = exports.rdau = exports.qudt = exports.quantitykind = void 0;
var acl_1 = require("./vocabularies/acl");
Object.defineProperty(exports, "acl", { enumerable: true, get: function () { return acl_1.strict; } });
var as_1 = require("./vocabularies/as");
Object.defineProperty(exports, "as", { enumerable: true, get: function () { return as_1.strict; } });
var bibo_1 = require("./vocabularies/bibo");
Object.defineProperty(exports, "bibo", { enumerable: true, get: function () { return bibo_1.strict; } });
var cc_1 = require("./vocabularies/cc");
Object.defineProperty(exports, "cc", { enumerable: true, get: function () { return cc_1.strict; } });
var cnt_1 = require("./vocabularies/cnt");
Object.defineProperty(exports, "cnt", { enumerable: true, get: function () { return cnt_1.strict; } });
var constant_1 = require("./vocabularies/constant");
Object.defineProperty(exports, "constant", { enumerable: true, get: function () { return constant_1.strict; } });
var csvw_1 = require("./vocabularies/csvw");
Object.defineProperty(exports, "csvw", { enumerable: true, get: function () { return csvw_1.strict; } });
var ctag_1 = require("./vocabularies/ctag");
Object.defineProperty(exports, "ctag", { enumerable: true, get: function () { return ctag_1.strict; } });
var dash_1 = require("./vocabularies/dash");
Object.defineProperty(exports, "dash", { enumerable: true, get: function () { return dash_1.strict; } });
var dbo_1 = require("./vocabularies/dbo");
Object.defineProperty(exports, "dbo", { enumerable: true, get: function () { return dbo_1.strict; } });
var dc11_1 = require("./vocabularies/dc11");
Object.defineProperty(exports, "dc11", { enumerable: true, get: function () { return dc11_1.strict; } });
var dcam_1 = require("./vocabularies/dcam");
Object.defineProperty(exports, "dcam", { enumerable: true, get: function () { return dcam_1.strict; } });
var dcat_1 = require("./vocabularies/dcat");
Object.defineProperty(exports, "dcat", { enumerable: true, get: function () { return dcat_1.strict; } });
var dcmitype_1 = require("./vocabularies/dcmitype");
Object.defineProperty(exports, "dcmitype", { enumerable: true, get: function () { return dcmitype_1.strict; } });
var dcterms_1 = require("./vocabularies/dcterms");
Object.defineProperty(exports, "dcterms", { enumerable: true, get: function () { return dcterms_1.strict; } });
var discipline_1 = require("./vocabularies/discipline");
Object.defineProperty(exports, "discipline", { enumerable: true, get: function () { return discipline_1.strict; } });
var doap_1 = require("./vocabularies/doap");
Object.defineProperty(exports, "doap", { enumerable: true, get: function () { return doap_1.strict; } });
var dpv_1 = require("./vocabularies/dpv");
Object.defineProperty(exports, "dpv", { enumerable: true, get: function () { return dpv_1.strict; } });
var dqv_1 = require("./vocabularies/dqv");
Object.defineProperty(exports, "dqv", { enumerable: true, get: function () { return dqv_1.strict; } });
var dtype_1 = require("./vocabularies/dtype");
Object.defineProperty(exports, "dtype", { enumerable: true, get: function () { return dtype_1.strict; } });
var duv_1 = require("./vocabularies/duv");
Object.defineProperty(exports, "duv", { enumerable: true, get: function () { return duv_1.strict; } });
var earl_1 = require("./vocabularies/earl");
Object.defineProperty(exports, "earl", { enumerable: true, get: function () { return earl_1.strict; } });
var ebucore_1 = require("./vocabularies/ebucore");
Object.defineProperty(exports, "ebucore", { enumerable: true, get: function () { return ebucore_1.strict; } });
var exif_1 = require("./vocabularies/exif");
Object.defineProperty(exports, "exif", { enumerable: true, get: function () { return exif_1.strict; } });
var foaf_1 = require("./vocabularies/foaf");
Object.defineProperty(exports, "foaf", { enumerable: true, get: function () { return foaf_1.strict; } });
var frbr_1 = require("./vocabularies/frbr");
Object.defineProperty(exports, "frbr", { enumerable: true, get: function () { return frbr_1.strict; } });
var geo_1 = require("./vocabularies/geo");
Object.defineProperty(exports, "geo", { enumerable: true, get: function () { return geo_1.strict; } });
var geof_1 = require("./vocabularies/geof");
Object.defineProperty(exports, "geof", { enumerable: true, get: function () { return geof_1.strict; } });
var geor_1 = require("./vocabularies/geor");
Object.defineProperty(exports, "geor", { enumerable: true, get: function () { return geor_1.strict; } });
var gml_1 = require("./vocabularies/gml");
Object.defineProperty(exports, "gml", { enumerable: true, get: function () { return gml_1.strict; } });
var gn_1 = require("./vocabularies/gn");
Object.defineProperty(exports, "gn", { enumerable: true, get: function () { return gn_1.strict; } });
var gr_1 = require("./vocabularies/gr");
Object.defineProperty(exports, "gr", { enumerable: true, get: function () { return gr_1.strict; } });
var grddl_1 = require("./vocabularies/grddl");
Object.defineProperty(exports, "grddl", { enumerable: true, get: function () { return grddl_1.strict; } });
var gs1_1 = require("./vocabularies/gs1");
Object.defineProperty(exports, "gs1", { enumerable: true, get: function () { return gs1_1.strict; } });
var gtfs_1 = require("./vocabularies/gtfs");
Object.defineProperty(exports, "gtfs", { enumerable: true, get: function () { return gtfs_1.strict; } });
var http_1 = require("./vocabularies/http");
Object.defineProperty(exports, "http", { enumerable: true, get: function () { return http_1.strict; } });
var hydra_1 = require("./vocabularies/hydra");
Object.defineProperty(exports, "hydra", { enumerable: true, get: function () { return hydra_1.strict; } });
var ical_1 = require("./vocabularies/ical");
Object.defineProperty(exports, "ical", { enumerable: true, get: function () { return ical_1.strict; } });
var ldp_1 = require("./vocabularies/ldp");
Object.defineProperty(exports, "ldp", { enumerable: true, get: function () { return ldp_1.strict; } });
var locn_1 = require("./vocabularies/locn");
Object.defineProperty(exports, "locn", { enumerable: true, get: function () { return locn_1.strict; } });
var lvont_1 = require("./vocabularies/lvont");
Object.defineProperty(exports, "lvont", { enumerable: true, get: function () { return lvont_1.strict; } });
var ma_1 = require("./vocabularies/ma");
Object.defineProperty(exports, "ma", { enumerable: true, get: function () { return ma_1.strict; } });
var oa_1 = require("./vocabularies/oa");
Object.defineProperty(exports, "oa", { enumerable: true, get: function () { return oa_1.strict; } });
var og_1 = require("./vocabularies/og");
Object.defineProperty(exports, "og", { enumerable: true, get: function () { return og_1.strict; } });
var org_1 = require("./vocabularies/org");
Object.defineProperty(exports, "org", { enumerable: true, get: function () { return org_1.strict; } });
var owl_1 = require("./vocabularies/owl");
Object.defineProperty(exports, "owl", { enumerable: true, get: function () { return owl_1.strict; } });
var prefix_1 = require("./vocabularies/prefix");
Object.defineProperty(exports, "prefix", { enumerable: true, get: function () { return prefix_1.strict; } });
var prov_1 = require("./vocabularies/prov");
Object.defineProperty(exports, "prov", { enumerable: true, get: function () { return prov_1.strict; } });
var qb_1 = require("./vocabularies/qb");
Object.defineProperty(exports, "qb", { enumerable: true, get: function () { return qb_1.strict; } });
var qkdv_1 = require("./vocabularies/qkdv");
Object.defineProperty(exports, "qkdv", { enumerable: true, get: function () { return qkdv_1.strict; } });
var quantitykind_1 = require("./vocabularies/quantitykind");
Object.defineProperty(exports, "quantitykind", { enumerable: true, get: function () { return quantitykind_1.strict; } });
var qudt_1 = require("./vocabularies/qudt");
Object.defineProperty(exports, "qudt", { enumerable: true, get: function () { return qudt_1.strict; } });
var rdau_1 = require("./vocabularies/rdau");
Object.defineProperty(exports, "rdau", { enumerable: true, get: function () { return rdau_1.strict; } });
var rdf_1 = require("./vocabularies/rdf");
Object.defineProperty(exports, "rdf", { enumerable: true, get: function () { return rdf_1.strict; } });
var rdfa_1 = require("./vocabularies/rdfa");
Object.defineProperty(exports, "rdfa", { enumerable: true, get: function () { return rdfa_1.strict; } });
var rdfs_1 = require("./vocabularies/rdfs");
Object.defineProperty(exports, "rdfs", { enumerable: true, get: function () { return rdfs_1.strict; } });
var rev_1 = require("./vocabularies/rev");
Object.defineProperty(exports, "rev", { enumerable: true, get: function () { return rev_1.strict; } });
var rico_1 = require("./vocabularies/rico");
Object.defineProperty(exports, "rico", { enumerable: true, get: function () { return rico_1.strict; } });
var rif_1 = require("./vocabularies/rif");
Object.defineProperty(exports, "rif", { enumerable: true, get: function () { return rif_1.strict; } });
var rr_1 = require("./vocabularies/rr");
Object.defineProperty(exports, "rr", { enumerable: true, get: function () { return rr_1.strict; } });
var rss_1 = require("./vocabularies/rss");
Object.defineProperty(exports, "rss", { enumerable: true, get: function () { return rss_1.strict; } });
var schema_1 = require("./vocabularies/schema");
Object.defineProperty(exports, "schema", { enumerable: true, get: function () { return schema_1.strict; } });
var sd_1 = require("./vocabularies/sd");
Object.defineProperty(exports, "sd", { enumerable: true, get: function () { return sd_1.strict; } });
var sdmx_1 = require("./vocabularies/sdmx");
Object.defineProperty(exports, "sdmx", { enumerable: true, get: function () { return sdmx_1.strict; } });
var sem_1 = require("./vocabularies/sem");
Object.defineProperty(exports, "sem", { enumerable: true, get: function () { return sem_1.strict; } });
var sf_1 = require("./vocabularies/sf");
Object.defineProperty(exports, "sf", { enumerable: true, get: function () { return sf_1.strict; } });
var sh_1 = require("./vocabularies/sh");
Object.defineProperty(exports, "sh", { enumerable: true, get: function () { return sh_1.strict; } });
var shex_1 = require("./vocabularies/shex");
Object.defineProperty(exports, "shex", { enumerable: true, get: function () { return shex_1.strict; } });
var sioc_1 = require("./vocabularies/sioc");
Object.defineProperty(exports, "sioc", { enumerable: true, get: function () { return sioc_1.strict; } });
var skos_1 = require("./vocabularies/skos");
Object.defineProperty(exports, "skos", { enumerable: true, get: function () { return skos_1.strict; } });
var skosxl_1 = require("./vocabularies/skosxl");
Object.defineProperty(exports, "skosxl", { enumerable: true, get: function () { return skosxl_1.strict; } });
var sosa_1 = require("./vocabularies/sosa");
Object.defineProperty(exports, "sosa", { enumerable: true, get: function () { return sosa_1.strict; } });
var sou_1 = require("./vocabularies/sou");
Object.defineProperty(exports, "sou", { enumerable: true, get: function () { return sou_1.strict; } });
var ssn_1 = require("./vocabularies/ssn");
Object.defineProperty(exports, "ssn", { enumerable: true, get: function () { return ssn_1.strict; } });
var test_1 = require("./vocabularies/test");
Object.defineProperty(exports, "test", { enumerable: true, get: function () { return test_1.strict; } });
var time_1 = require("./vocabularies/time");
Object.defineProperty(exports, "time", { enumerable: true, get: function () { return time_1.strict; } });
var unit_1 = require("./vocabularies/unit");
Object.defineProperty(exports, "unit", { enumerable: true, get: function () { return unit_1.strict; } });
var v_1 = require("./vocabularies/v");
Object.defineProperty(exports, "v", { enumerable: true, get: function () { return v_1.strict; } });
var vaem_1 = require("./vocabularies/vaem");
Object.defineProperty(exports, "vaem", { enumerable: true, get: function () { return vaem_1.strict; } });
var vann_1 = require("./vocabularies/vann");
Object.defineProperty(exports, "vann", { enumerable: true, get: function () { return vann_1.strict; } });
var vcard_1 = require("./vocabularies/vcard");
Object.defineProperty(exports, "vcard", { enumerable: true, get: function () { return vcard_1.strict; } });
var void_1 = require("./vocabularies/void");
Object.defineProperty(exports, "_void", { enumerable: true, get: function () { return void_1.strict; } });
var vs_1 = require("./vocabularies/vs");
Object.defineProperty(exports, "vs", { enumerable: true, get: function () { return vs_1.strict; } });
var wdr_1 = require("./vocabularies/wdr");
Object.defineProperty(exports, "wdr", { enumerable: true, get: function () { return wdr_1.strict; } });
var wdrs_1 = require("./vocabularies/wdrs");
Object.defineProperty(exports, "wdrs", { enumerable: true, get: function () { return wdrs_1.strict; } });
var wgs_1 = require("./vocabularies/wgs");
Object.defineProperty(exports, "wgs", { enumerable: true, get: function () { return wgs_1.strict; } });
var xhv_1 = require("./vocabularies/xhv");
Object.defineProperty(exports, "xhv", { enumerable: true, get: function () { return xhv_1.strict; } });
var xkos_1 = require("./vocabularies/xkos");
Object.defineProperty(exports, "xkos", { enumerable: true, get: function () { return xkos_1.strict; } });
var xml_1 = require("./vocabularies/xml");
Object.defineProperty(exports, "xml", { enumerable: true, get: function () { return xml_1.strict; } });
var xsd_1 = require("./vocabularies/xsd");
Object.defineProperty(exports, "xsd", { enumerable: true, get: function () { return xsd_1.strict; } });
