// @tpluscode/rdf-ns-builders/index.mjs
export { strict as acl } from "./vocabularies/acl.mjs";
export { strict as as } from "./vocabularies/as.mjs";
export { strict as bibo } from "./vocabularies/bibo.mjs";
export { strict as cc } from "./vocabularies/cc.mjs";
export { strict as cnt } from "./vocabularies/cnt.mjs";
export { strict as constant } from "./vocabularies/constant.mjs";
export { strict as csvw } from "./vocabularies/csvw.mjs";
export { strict as ctag } from "./vocabularies/ctag.mjs";
export { strict as dash } from "./vocabularies/dash.mjs";
export { strict as dbo } from "./vocabularies/dbo.mjs";
export { strict as dc11 } from "./vocabularies/dc11.mjs";
export { strict as dcam } from "./vocabularies/dcam.mjs";
export { strict as dcat } from "./vocabularies/dcat.mjs";
export { strict as dcmitype } from "./vocabularies/dcmitype.mjs";
export { strict as dcterms } from "./vocabularies/dcterms.mjs";
export { strict as discipline } from "./vocabularies/discipline.mjs";
export { strict as doap } from "./vocabularies/doap.mjs";
export { strict as dpv } from "./vocabularies/dpv.mjs";
export { strict as dqv } from "./vocabularies/dqv.mjs";
export { strict as dtype } from "./vocabularies/dtype.mjs";
export { strict as duv } from "./vocabularies/duv.mjs";
export { strict as earl } from "./vocabularies/earl.mjs";
export { strict as ebucore } from "./vocabularies/ebucore.mjs";
export { strict as exif } from "./vocabularies/exif.mjs";
export { strict as foaf } from "./vocabularies/foaf.mjs";
export { strict as frbr } from "./vocabularies/frbr.mjs";
export { strict as geo } from "./vocabularies/geo.mjs";
export { strict as geof } from "./vocabularies/geof.mjs";
export { strict as geor } from "./vocabularies/geor.mjs";
export { strict as gml } from "./vocabularies/gml.mjs";
export { strict as gn } from "./vocabularies/gn.mjs";
export { strict as gr } from "./vocabularies/gr.mjs";
export { strict as grddl } from "./vocabularies/grddl.mjs";
export { strict as gs1 } from "./vocabularies/gs1.mjs";
export { strict as gtfs } from "./vocabularies/gtfs.mjs";
export { strict as http } from "./vocabularies/http.mjs";
export { strict as hydra } from "./vocabularies/hydra.mjs";
export { strict as ical } from "./vocabularies/ical.mjs";
export { strict as ldp } from "./vocabularies/ldp.mjs";
export { strict as locn } from "./vocabularies/locn.mjs";
export { strict as lvont } from "./vocabularies/lvont.mjs";
export { strict as ma } from "./vocabularies/ma.mjs";
export { strict as oa } from "./vocabularies/oa.mjs";
export { strict as og } from "./vocabularies/og.mjs";
export { strict as org } from "./vocabularies/org.mjs";
export { strict as owl } from "./vocabularies/owl.mjs";
export { strict as prefix } from "./vocabularies/prefix.mjs";
export { strict as prov } from "./vocabularies/prov.mjs";
export { strict as qb } from "./vocabularies/qb.mjs";
export { strict as qkdv } from "./vocabularies/qkdv.mjs";
export { strict as quantitykind } from "./vocabularies/quantitykind.mjs";
export { strict as qudt } from "./vocabularies/qudt.mjs";
export { strict as rdau } from "./vocabularies/rdau.mjs";
export { strict as rdf } from "./vocabularies/rdf.mjs";
export { strict as rdfa } from "./vocabularies/rdfa.mjs";
export { strict as rdfs } from "./vocabularies/rdfs.mjs";
export { strict as rev } from "./vocabularies/rev.mjs";
export { strict as rico } from "./vocabularies/rico.mjs";
export { strict as rif } from "./vocabularies/rif.mjs";
export { strict as rr } from "./vocabularies/rr.mjs";
export { strict as rss } from "./vocabularies/rss.mjs";
export { strict as schema } from "./vocabularies/schema.mjs";
export { strict as sd } from "./vocabularies/sd.mjs";
export { strict as sdmx } from "./vocabularies/sdmx.mjs";
export { strict as sem } from "./vocabularies/sem.mjs";
export { strict as sf } from "./vocabularies/sf.mjs";
export { strict as sh } from "./vocabularies/sh.mjs";
export { strict as shex } from "./vocabularies/shex.mjs";
export { strict as sioc } from "./vocabularies/sioc.mjs";
export { strict as skos } from "./vocabularies/skos.mjs";
export { strict as skosxl } from "./vocabularies/skosxl.mjs";
export { strict as sosa } from "./vocabularies/sosa.mjs";
export { strict as sou } from "./vocabularies/sou.mjs";
export { strict as ssn } from "./vocabularies/ssn.mjs";
export { strict as test } from "./vocabularies/test.mjs";
export { strict as time } from "./vocabularies/time.mjs";
export { strict as unit } from "./vocabularies/unit.mjs";
export { strict as v } from "./vocabularies/v.mjs";
export { strict as vaem } from "./vocabularies/vaem.mjs";
export { strict as vann } from "./vocabularies/vann.mjs";
export { strict as vcard } from "./vocabularies/vcard.mjs";
export { strict as _void } from "./vocabularies/void.mjs";
export { strict as vs } from "./vocabularies/vs.mjs";
export { strict as wdr } from "./vocabularies/wdr.mjs";
export { strict as wdrs } from "./vocabularies/wdrs.mjs";
export { strict as wgs } from "./vocabularies/wgs.mjs";
export { strict as xhv } from "./vocabularies/xhv.mjs";
export { strict as xkos } from "./vocabularies/xkos.mjs";
export { strict as xml } from "./vocabularies/xml.mjs";
export { strict as xsd } from "./vocabularies/xsd.mjs";
