// @tpluscode/rdf-ns-builders/loose.mjs
export { loose as acl } from "./vocabularies/acl.mjs";
export { loose as as } from "./vocabularies/as.mjs";
export { loose as bibo } from "./vocabularies/bibo.mjs";
export { loose as cc } from "./vocabularies/cc.mjs";
export { loose as cnt } from "./vocabularies/cnt.mjs";
export { loose as constant } from "./vocabularies/constant.mjs";
export { loose as csvw } from "./vocabularies/csvw.mjs";
export { loose as ctag } from "./vocabularies/ctag.mjs";
export { loose as dash } from "./vocabularies/dash.mjs";
export { loose as dbo } from "./vocabularies/dbo.mjs";
export { loose as dc11 } from "./vocabularies/dc11.mjs";
export { loose as dcam } from "./vocabularies/dcam.mjs";
export { loose as dcat } from "./vocabularies/dcat.mjs";
export { loose as dcmitype } from "./vocabularies/dcmitype.mjs";
export { loose as dcterms } from "./vocabularies/dcterms.mjs";
export { loose as discipline } from "./vocabularies/discipline.mjs";
export { loose as doap } from "./vocabularies/doap.mjs";
export { loose as dpv } from "./vocabularies/dpv.mjs";
export { loose as dqv } from "./vocabularies/dqv.mjs";
export { loose as dtype } from "./vocabularies/dtype.mjs";
export { loose as duv } from "./vocabularies/duv.mjs";
export { loose as earl } from "./vocabularies/earl.mjs";
export { loose as ebucore } from "./vocabularies/ebucore.mjs";
export { loose as exif } from "./vocabularies/exif.mjs";
export { loose as foaf } from "./vocabularies/foaf.mjs";
export { loose as frbr } from "./vocabularies/frbr.mjs";
export { loose as geo } from "./vocabularies/geo.mjs";
export { loose as geof } from "./vocabularies/geof.mjs";
export { loose as geor } from "./vocabularies/geor.mjs";
export { loose as gml } from "./vocabularies/gml.mjs";
export { loose as gn } from "./vocabularies/gn.mjs";
export { loose as gr } from "./vocabularies/gr.mjs";
export { loose as grddl } from "./vocabularies/grddl.mjs";
export { loose as gs1 } from "./vocabularies/gs1.mjs";
export { loose as gtfs } from "./vocabularies/gtfs.mjs";
export { loose as http } from "./vocabularies/http.mjs";
export { loose as hydra } from "./vocabularies/hydra.mjs";
export { loose as ical } from "./vocabularies/ical.mjs";
export { loose as ldp } from "./vocabularies/ldp.mjs";
export { loose as locn } from "./vocabularies/locn.mjs";
export { loose as lvont } from "./vocabularies/lvont.mjs";
export { loose as ma } from "./vocabularies/ma.mjs";
export { loose as oa } from "./vocabularies/oa.mjs";
export { loose as og } from "./vocabularies/og.mjs";
export { loose as org } from "./vocabularies/org.mjs";
export { loose as owl } from "./vocabularies/owl.mjs";
export { loose as prefix } from "./vocabularies/prefix.mjs";
export { loose as prov } from "./vocabularies/prov.mjs";
export { loose as qb } from "./vocabularies/qb.mjs";
export { loose as qkdv } from "./vocabularies/qkdv.mjs";
export { loose as quantitykind } from "./vocabularies/quantitykind.mjs";
export { loose as qudt } from "./vocabularies/qudt.mjs";
export { loose as rdau } from "./vocabularies/rdau.mjs";
export { loose as rdf } from "./vocabularies/rdf.mjs";
export { loose as rdfa } from "./vocabularies/rdfa.mjs";
export { loose as rdfs } from "./vocabularies/rdfs.mjs";
export { loose as rev } from "./vocabularies/rev.mjs";
export { loose as rico } from "./vocabularies/rico.mjs";
export { loose as rif } from "./vocabularies/rif.mjs";
export { loose as rr } from "./vocabularies/rr.mjs";
export { loose as rss } from "./vocabularies/rss.mjs";
export { loose as schema } from "./vocabularies/schema.mjs";
export { loose as sd } from "./vocabularies/sd.mjs";
export { loose as sdmx } from "./vocabularies/sdmx.mjs";
export { loose as sem } from "./vocabularies/sem.mjs";
export { loose as sf } from "./vocabularies/sf.mjs";
export { loose as sh } from "./vocabularies/sh.mjs";
export { loose as shex } from "./vocabularies/shex.mjs";
export { loose as sioc } from "./vocabularies/sioc.mjs";
export { loose as skos } from "./vocabularies/skos.mjs";
export { loose as skosxl } from "./vocabularies/skosxl.mjs";
export { loose as sosa } from "./vocabularies/sosa.mjs";
export { loose as sou } from "./vocabularies/sou.mjs";
export { loose as ssn } from "./vocabularies/ssn.mjs";
export { loose as test } from "./vocabularies/test.mjs";
export { loose as time } from "./vocabularies/time.mjs";
export { loose as unit } from "./vocabularies/unit.mjs";
export { loose as v } from "./vocabularies/v.mjs";
export { loose as vaem } from "./vocabularies/vaem.mjs";
export { loose as vann } from "./vocabularies/vann.mjs";
export { loose as vcard } from "./vocabularies/vcard.mjs";
export { loose as _void } from "./vocabularies/void.mjs";
export { loose as vs } from "./vocabularies/vs.mjs";
export { loose as wdr } from "./vocabularies/wdr.mjs";
export { loose as wdrs } from "./vocabularies/wdrs.mjs";
export { loose as wgs } from "./vocabularies/wgs.mjs";
export { loose as xhv } from "./vocabularies/xhv.mjs";
export { loose as xkos } from "./vocabularies/xkos.mjs";
export { loose as xml } from "./vocabularies/xml.mjs";
export { loose as xsd } from "./vocabularies/xsd.mjs";
