// @zazuko/rdf-vocabularies/datasets/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var acl_1 = require("./acl");
exports.acl = acl_1.default;
var as_1 = require("./as");
exports.as = as_1.default;
var bibo_1 = require("./bibo");
exports.bibo = bibo_1.default;
var cc_1 = require("./cc");
exports.cc = cc_1.default;
var cnt_1 = require("./cnt");
exports.cnt = cnt_1.default;
var constant_1 = require("./constant");
exports.constant = constant_1.default;
var crm_1 = require("./crm");
exports.crm = crm_1.default;
var csvw_1 = require("./csvw");
exports.csvw = csvw_1.default;
var dash_1 = require("./dash");
exports.dash = dash_1.default;
var dbo_1 = require("./dbo");
exports.dbo = dbo_1.default;
var dc11_1 = require("./dc11");
exports.dc11 = dc11_1.default;
var dcam_1 = require("./dcam");
exports.dcam = dcam_1.default;
var dcat_1 = require("./dcat");
exports.dcat = dcat_1.default;
var dcmitype_1 = require("./dcmitype");
exports.dcmitype = dcmitype_1.default;
var dcterms_1 = require("./dcterms");
exports.dcterms = dcterms_1.default;
var discipline_1 = require("./discipline");
exports.discipline = discipline_1.default;
var doap_1 = require("./doap");
exports.doap = doap_1.default;
var dpv_1 = require("./dpv");
exports.dpv = dpv_1.default;
var dqv_1 = require("./dqv");
exports.dqv = dqv_1.default;
var dtype_1 = require("./dtype");
exports.dtype = dtype_1.default;
var duv_1 = require("./duv");
exports.duv = duv_1.default;
var earl_1 = require("./earl");
exports.earl = earl_1.default;
var ebucore_1 = require("./ebucore");
exports.ebucore = ebucore_1.default;
var exif_1 = require("./exif");
exports.exif = exif_1.default;
var foaf_1 = require("./foaf");
exports.foaf = foaf_1.default;
var frbr_1 = require("./frbr");
exports.frbr = frbr_1.default;
var geo_1 = require("./geo");
exports.geo = geo_1.default;
var geof_1 = require("./geof");
exports.geof = geof_1.default;
var geor_1 = require("./geor");
exports.geor = geor_1.default;
var gn_1 = require("./gn");
exports.gn = gn_1.default;
var gr_1 = require("./gr");
exports.gr = gr_1.default;
var grddl_1 = require("./grddl");
exports.grddl = grddl_1.default;
var gs1_1 = require("./gs1");
exports.gs1 = gs1_1.default;
var gtfs_1 = require("./gtfs");
exports.gtfs = gtfs_1.default;
var http_1 = require("./http");
exports.http = http_1.default;
var hydra_1 = require("./hydra");
exports.hydra = hydra_1.default;
var ical_1 = require("./ical");
exports.ical = ical_1.default;
var ldp_1 = require("./ldp");
exports.ldp = ldp_1.default;
var locn_1 = require("./locn");
exports.locn = locn_1.default;
var lvont_1 = require("./lvont");
exports.lvont = lvont_1.default;
var ma_1 = require("./ma");
exports.ma = ma_1.default;
var mads_1 = require("./mads");
exports.mads = mads_1.default;
var oa_1 = require("./oa");
exports.oa = oa_1.default;
var og_1 = require("./og");
exports.og = og_1.default;
var org_1 = require("./org");
exports.org = org_1.default;
var owl_1 = require("./owl");
exports.owl = owl_1.default;
var prefix_1 = require("./prefix");
exports.prefix = prefix_1.default;
var prov_1 = require("./prov");
exports.prov = prov_1.default;
var qb_1 = require("./qb");
exports.qb = qb_1.default;
var qkdv_1 = require("./qkdv");
exports.qkdv = qkdv_1.default;
var quantitykind_1 = require("./quantitykind");
exports.quantitykind = quantitykind_1.default;
var qudt_1 = require("./qudt");
exports.qudt = qudt_1.default;
var rdau_1 = require("./rdau");
exports.rdau = rdau_1.default;
var rdf_1 = require("./rdf");
exports.rdf = rdf_1.default;
var rdfa_1 = require("./rdfa");
exports.rdfa = rdfa_1.default;
var rdfs_1 = require("./rdfs");
exports.rdfs = rdfs_1.default;
var rico_1 = require("./rico");
exports.rico = rico_1.default;
var rr_1 = require("./rr");
exports.rr = rr_1.default;
var rss_1 = require("./rss");
exports.rss = rss_1.default;
var schema_1 = require("./schema");
exports.schema = schema_1.default;
var sd_1 = require("./sd");
exports.sd = sd_1.default;
var sdmx_1 = require("./sdmx");
exports.sdmx = sdmx_1.default;
var sem_1 = require("./sem");
exports.sem = sem_1.default;
var sh_1 = require("./sh");
exports.sh = sh_1.default;
var shex_1 = require("./shex");
exports.shex = shex_1.default;
var sioc_1 = require("./sioc");
exports.sioc = sioc_1.default;
var skos_1 = require("./skos");
exports.skos = skos_1.default;
var skosxl_1 = require("./skosxl");
exports.skosxl = skosxl_1.default;
var sosa_1 = require("./sosa");
exports.sosa = sosa_1.default;
var sou_1 = require("./sou");
exports.sou = sou_1.default;
var ssn_1 = require("./ssn");
exports.ssn = ssn_1.default;
var test_1 = require("./test");
exports.test = test_1.default;
var time_1 = require("./time");
exports.time = time_1.default;
var unit_1 = require("./unit");
exports.unit = unit_1.default;
var vaem_1 = require("./vaem");
exports.vaem = vaem_1.default;
var vann_1 = require("./vann");
exports.vann = vann_1.default;
var vcard_1 = require("./vcard");
exports.vcard = vcard_1.default;
var void_1 = require("./void");
exports.void = void_1.default;
var vs_1 = require("./vs");
exports.vs = vs_1.default;
var wgs_1 = require("./wgs");
exports.wgs = wgs_1.default;
var xhv_1 = require("./xhv");
exports.xhv = xhv_1.default;
var xkos_1 = require("./xkos");
exports.xkos = xkos_1.default;
var xsd_1 = require("./xsd");
exports.xsd = xsd_1.default;
