// @zazuko/rdf-vocabularies/datasets/index.mjs
export { default as acl } from "./acl.mjs";
export { default as as } from "./as.mjs";
export { default as bibo } from "./bibo.mjs";
export { default as cc } from "./cc.mjs";
export { default as cnt } from "./cnt.mjs";
export { default as constant } from "./constant.mjs";
export { default as crm } from "./crm.mjs";
export { default as csvw } from "./csvw.mjs";
export { default as dash } from "./dash.mjs";
export { default as dbo } from "./dbo.mjs";
export { default as dc11 } from "./dc11.mjs";
export { default as dcam } from "./dcam.mjs";
export { default as dcat } from "./dcat.mjs";
export { default as dcmitype } from "./dcmitype.mjs";
export { default as dcterms } from "./dcterms.mjs";
export { default as discipline } from "./discipline.mjs";
export { default as doap } from "./doap.mjs";
export { default as dpv } from "./dpv.mjs";
export { default as dqv } from "./dqv.mjs";
export { default as dtype } from "./dtype.mjs";
export { default as duv } from "./duv.mjs";
export { default as earl } from "./earl.mjs";
export { default as ebucore } from "./ebucore.mjs";
export { default as exif } from "./exif.mjs";
export { default as foaf } from "./foaf.mjs";
export { default as frbr } from "./frbr.mjs";
export { default as geo } from "./geo.mjs";
export { default as geof } from "./geof.mjs";
export { default as geor } from "./geor.mjs";
export { default as gn } from "./gn.mjs";
export { default as gr } from "./gr.mjs";
export { default as grddl } from "./grddl.mjs";
export { default as gs1 } from "./gs1.mjs";
export { default as gtfs } from "./gtfs.mjs";
export { default as http } from "./http.mjs";
export { default as hydra } from "./hydra.mjs";
export { default as ical } from "./ical.mjs";
export { default as ldp } from "./ldp.mjs";
export { default as locn } from "./locn.mjs";
export { default as lvont } from "./lvont.mjs";
export { default as ma } from "./ma.mjs";
export { default as mads } from "./mads.mjs";
export { default as oa } from "./oa.mjs";
export { default as og } from "./og.mjs";
export { default as org } from "./org.mjs";
export { default as owl } from "./owl.mjs";
export { default as prefix } from "./prefix.mjs";
export { default as prov } from "./prov.mjs";
export { default as qb } from "./qb.mjs";
export { default as qkdv } from "./qkdv.mjs";
export { default as quantitykind } from "./quantitykind.mjs";
export { default as qudt } from "./qudt.mjs";
export { default as rdau } from "./rdau.mjs";
export { default as rdf } from "./rdf.mjs";
export { default as rdfa } from "./rdfa.mjs";
export { default as rdfs } from "./rdfs.mjs";
export { default as rico } from "./rico.mjs";
export { default as rr } from "./rr.mjs";
export { default as rss } from "./rss.mjs";
export { default as schema } from "./schema.mjs";
export { default as sd } from "./sd.mjs";
export { default as sdmx } from "./sdmx.mjs";
export { default as sem } from "./sem.mjs";
export { default as sh } from "./sh.mjs";
export { default as shex } from "./shex.mjs";
export { default as sioc } from "./sioc.mjs";
export { default as skos } from "./skos.mjs";
export { default as skosxl } from "./skosxl.mjs";
export { default as sosa } from "./sosa.mjs";
export { default as sou } from "./sou.mjs";
export { default as ssn } from "./ssn.mjs";
export { default as test } from "./test.mjs";
export { default as time } from "./time.mjs";
export { default as unit } from "./unit.mjs";
export { default as vaem } from "./vaem.mjs";
export { default as vann } from "./vann.mjs";
export { default as vcard } from "./vcard.mjs";
export { default as void } from "./void.mjs";
export { default as vs } from "./vs.mjs";
export { default as wgs } from "./wgs.mjs";
export { default as xhv } from "./xhv.mjs";
export { default as xkos } from "./xkos.mjs";
export { default as xsd } from "./xsd.mjs";
