// @treecg/tree-metadata-extraction/dist/util/Util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelationType = void 0;
var RelationType;
(function (RelationType) {
    RelationType["Relation"] = "https://w3id.org/tree#Relation";
    RelationType["PrefixRelation"] = "https://w3id.org/tree#PrefixRelation";
    RelationType["SubstringRelation"] = "https://w3id.org/tree#SubstringRelation";
    RelationType["GreaterThanRelation"] = "https://w3id.org/tree#GreaterThanRelation";
    RelationType["GreaterThanOrEqualToRelation"] = "https://w3id.org/tree#GreaterThanOrEqualToRelation";
    RelationType["LessThanRelation"] = "https://w3id.org/tree#LessThanRelation";
    RelationType["LessThanOrEqualToRelation"] = "https://w3id.org/tree#LessThanOrEqualToRelation";
    RelationType["EqualThanRelation"] = "https://w3id.org/tree#EqualThanRelation";
    RelationType["GeospatiallyContainsRelation"] = "https://w3id.org/tree#GeospatiallyContainsRelation";
    RelationType["InBetweenRelation"] = "https://w3id.org/tree#InBetweenRelation";
})(RelationType = exports.RelationType || (exports.RelationType = {}));
