// @tpluscode/rdf-string/lib/ntriples.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ntriples = exports.NTriplesStrategy = void 0;
const data_model_1 = __importDefault(require("@rdf-esm/data-model"));
const nquads_1 = require("./nquads");
const TemplateResult_1 = require("./TemplateResult");
class NTriplesStrategy extends nquads_1.NQuadsStrategy {
    evaluateDataset(dataset, options) {
        const singleGraph = dataset.match(null, null, null, options.graph);
        return super.evaluateDataset(singleGraph, options);
    }
    evaluateQuad(quad, options) {
        if (!options.graph.equals(quad.graph)) {
            return {
                value: '',
                prefixes: [],
            };
        }
        return super.evaluateQuad(data_model_1.default.triple(quad.subject, quad.predicate, quad.object), options);
    }
}
exports.NTriplesStrategy = NTriplesStrategy;
exports.ntriples = (strings, ...values) => new TemplateResult_1.TemplateResult({
    strings,
    values,
    tag: exports.ntriples,
    strategy: new NTriplesStrategy(),
    defaultOptions: {
        graph: data_model_1.default.defaultGraph(),
        sortGraphs: false,
    },
});
