// @comunica/actor-rdf-resolve-quad-pattern-hypermedia/lib/MediatedQuadSource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediatedQuadSource = void 0;
const MediatedLinkedRdfSourcesAsyncRdfIterator_1 = require("./MediatedLinkedRdfSourcesAsyncRdfIterator");
const StreamingStoreMetadata_1 = require("./StreamingStoreMetadata");
/**
 * A lazy quad source that creates {@link MediatedLinkedRdfSourcesAsyncRdfIterator} instances when matching quads.
 *
 * @see MediatedLinkedRdfSourcesAsyncRdfIterator
 */
class MediatedQuadSource {
    constructor(cacheSize, context, firstUrl, forceSourceType, maxIterators, aggregateStore, mediators) {
        this.cacheSize = cacheSize;
        this.context = context;
        this.firstUrl = firstUrl;
        this.forceSourceType = forceSourceType;
        this.maxIterators = maxIterators;
        this.aggregatedStore = aggregateStore ? new StreamingStoreMetadata_1.StreamingStoreMetadata() : undefined;
        this.mediators = mediators;
    }
    static nullifyVariables(term) {
        return !term || term.termType === 'Variable' ? undefined : term;
    }
    match(subject, predicate, object, graph) {
        if (this.sourcesState && this.aggregatedStore) {
            return this.aggregatedStore.match(MediatedQuadSource.nullifyVariables(subject), MediatedQuadSource.nullifyVariables(predicate), MediatedQuadSource.nullifyVariables(object), MediatedQuadSource.nullifyVariables(graph));
        }
        const it = new MediatedLinkedRdfSourcesAsyncRdfIterator_1.MediatedLinkedRdfSourcesAsyncRdfIterator(this.cacheSize, this.context, this.forceSourceType, subject, predicate, object, graph, this.firstUrl, this.maxIterators, this.aggregatedStore, this.mediators);
        if (!this.sourcesState) {
            it.setSourcesState();
            this.sourcesState = it.sourcesState;
        }
        else {
            it.setSourcesState(this.sourcesState);
        }
        return it;
    }
}
exports.MediatedQuadSource = MediatedQuadSource;
